core/drivers/imx_rngb.c

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * (c) 2021 Jorge Ramirez <jorge@foundries.io>, Foundries Ltd.
+ */
+
+#include <arm.h>
+#include <crypto/crypto.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee/tee_cryp_utl.h>
+#include <trace.h>
+#include <util.h>
+
+#define RNGB_COMMAND			0x0004
+#define RNGB_CONTROL			0x0008
+#define RNGB_STATUS			0x000C
+#define RNGB_ERROR			0x0010
+#define RNGB_FIFO			0x0014
+
+#define RNGB_CMD_CLR_ERR		0x00000020
+#define RNGB_CMD_CLR_INT		0x00000010
+#define RNGB_CMD_SEED			0x00000002
+#define RNGB_CMD_SELF_TEST		0x00000001
+
+#define RNGB_CTRL_MASK_ERROR		0x00000040
+#define RNGB_CTRL_MASK_DONE		0x00000020
+
+#define RNGB_STATUS_ERROR		0x00010000
+#define RNGB_STATUS_FIFO_LEVEL_MASK	0x00000f00
+#define RNGB_STATUS_FIFO_LEVEL_SHIFT	8
+#define RNGB_STATUS_SEED_DONE		0x00000020
+#define RNGB_STATUS_ST_DONE		0x00000010
+
+#define RNGB_ERROR_STATUS_STAT_ERR	0x00000008
+
+#define RNGB_TIMEOUT			1000000
+
+static struct imx_rng {
+	struct io_pa_va base;
+	uint32_t error;
+} rngb = {
+#if defined(RNGB_BASE)
+	.base.pa = RNGB_BASE,
+#else
+#error IMX_RNGB driver not supported on this platform
+#endif
+};
+
+static void wait_for_irq(struct imx_rng *rng)
+{
+	uint64_t tref = timeout_init_us(RNGB_TIMEOUT);
+	uint32_t status = 0;
+
+	while (!(status & (RNGB_STATUS_SEED_DONE | RNGB_STATUS_ST_DONE))) {
+		rng->error = io_read32(rng->base.va + RNGB_ERROR);
+		status = io_read32(rng->base.va + RNGB_STATUS);
+
+		if (timeout_elapsed(tref))
+			panic();
+	}
+}
+
+static void irq_clear(struct imx_rng *rng)
+{
+	uint32_t ctrl = 0;
+	uint32_t cmd = 0;
+
+	ctrl = io_read32(rng->base.va + RNGB_CONTROL);
+	ctrl |= RNGB_CTRL_MASK_DONE | RNGB_CTRL_MASK_ERROR;
+	io_write32(rng->base.va + RNGB_CONTROL, ctrl);
+
+	cmd = io_read32(rng->base.va + RNGB_COMMAND);
+	cmd |= RNGB_CMD_CLR_INT | RNGB_CMD_CLR_ERR;
+	io_write32(rng->base.va + RNGB_COMMAND, cmd);
+}
+
+static void irq_unmask(struct imx_rng *rng)
+{
+	uint32_t ctrl = 0;
+
+	ctrl = io_read32(rng->base.va + RNGB_CONTROL);
+	ctrl &= ~(RNGB_CTRL_MASK_DONE | RNGB_CTRL_MASK_ERROR);
+	io_write32(rng->base.va + RNGB_CONTROL, ctrl);
+}
+
+static void rng_seed(struct imx_rng *rng)
```

Comment:
Should we use automatic re-seeding from hardware so that random numbers would be readily available when needed?

Reply:
yeah, that seems a better mode.